x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/aggs/InferencePipelineAggregationBuilder.java

```diff
@@ -0,0 +1,214 @@
+        modelId = in.readString();
+        bucketPathMap = in.readMap(StreamInput::readString, StreamInput::readString);
+        inferenceConfig = in.readOptionalNamedWriteable(InferenceConfigUpdate.class);
+        this.modelLoadingService = modelLoadingService;
+    }
+
+    void setModelId(String modelId) {
+        this.modelId = modelId;
+    }
+
+    void setInferenceConfig(InferenceConfigUpdate inferenceConfig) {
+        this.inferenceConfig = inferenceConfig;
+    }
+
+    @Override
+    protected void validate(ValidationContext context) {
+        context.validateHasParent(NAME, name);
+        if (modelId == null) {
+            context.addValidationError("[model_id] must be set");
+        }
+
+        if (inferenceConfig != null) {
+            // error if the results field is set and not equal to the only acceptable value
+            String resultsField = inferenceConfig.getResultsField();
+            if (Strings.isNullOrEmpty(resultsField) == false && AGGREGATIONS_RESULTS_FIELD.equals(resultsField) == false) {
+                context.addValidationError("setting option [" + ClassificationConfig.RESULTS_FIELD.getPreferredName()
+                    + "] to [" + resultsField + "] is not valid for inference aggregations");
+            }
+
+            if (inferenceConfig instanceof ClassificationConfigUpdate) {
+                ClassificationConfigUpdate classUpdate = (ClassificationConfigUpdate)inferenceConfig;
+
+                // error if the top classes result field is set and not equal to the only acceptable value
+                String topClassesField = classUpdate.getTopClassesResultsField();
+                if (Strings.isNullOrEmpty(topClassesField) == false &&
+                    ClassificationConfig.DEFAULT_TOP_CLASSES_RESULTS_FIELD.equals(topClassesField) == false) {
+                    context.addValidationError("setting option [" + ClassificationConfig.DEFAULT_TOP_CLASSES_RESULTS_FIELD
+                        + "] to [" + topClassesField + "] is not valid for inference aggregations");
+                }
+            }
+        }
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(modelId);
+        out.writeMap(bucketPathMap, StreamOutput::writeString, StreamOutput::writeString);
+        out.writeOptionalNamedWriteable(inferenceConfig);
+    }
+
+    @Override
+    protected PipelineAggregator createInternal(Map<String, Object> metaData) {
+
+        SetOnce<LocalModel> model = new SetOnce<>();
+        SetOnce<Exception> error = new SetOnce<>();
+        CountDownLatch latch = new CountDownLatch(1);
+        ActionListener<LocalModel> listener = new LatchedActionListener<>(
+            ActionListener.wrap(model::set, error::set), latch);
+
+        modelLoadingService.get().getModelForSearch(modelId, listener);
+        try {
+            // TODO Avoid the blocking wait
+            latch.await();
+        } catch (InterruptedException e) {
```

Comment:
I'm ok to merge this, but I'll work immediately on replacing it with a rewrite. I don't think we can release this code.

Reply:
I opened #58878 which should make this a thing we don't need to do any more. I've not made sure the tests pass with that yet, but I think it'll work soon enough.

Reply:
Thanks for the review